async/source.go

```diff
@@ -32,59 +56,71 @@ type MessageSource interface {
 // NewMessageSource returns a new insurance asynchronous message source, given
 // an AsyncMessageSource. When Close is called on the AsyncMessageSource,
 // this is also propogated to the underlying AsyncMessageSource.
-func NewMessageSource(ams substrate.AsyncMessageSource) MessageSource {
-	return &messageSourceAdapter{
-		ams,
+func NewMessageSource(source substrate.AsyncMessageSource, opts ...MessageSourceOption) MessageSource {
+	msa := &messageSourceAdapter{
+		source:    source,
+		consumers: 1,
+	}
+
+	for _, opt := range opts {
+		opt(msa)
 	}
+
+	return msa
 }
 
 type messageSourceAdapter struct {
-	ac substrate.AsyncMessageSource
+	source    substrate.AsyncMessageSource
+	msgBuffer int
+	ackBuffer int
+	consumers int
```

Comment:
Perhaps these should be `uint` or the `WithXXX` should have some sort of validation baked in? -1 consumers or `ackBuffer` doesn't sound healthy! :smile: